File: wakakusa/demo/lookup.rb
# A Hash on the app: reading with a fallback, asking whether a key is
# there, and adding one while the window is open.
require "wakakusa"

class Lookup
  def initialize
    @prices = { "apple" => 120, "banana" => 80 }
    @picked = 0
    @label = "none"
  end

  def pick_apple
    @picked = @prices.fetch("apple", -1)
    @label = @prices.key?("cherry") ? "cherry known" : "no cherry"
  end

  def add_cherry
    @prices["cherry"] = 200
    @picked = @prices.fetch("cherry", -1)
    @label = "cherry known" if @prices.key?("cherry")
  end

  def view
    column(spacing: 8.0, padding: 12.0) {
      text "picked=#{@picked} n=#{@prices.length} #{@label}"
      text "apple costs #{@prices.fetch("apple", -1)} right now", size: 12.0
      row(spacing: 6.0) {
        button("apple") { pick_apple }
        button("cherry") { add_cherry }
        button("miss") { @picked = @prices.fetch("durian", -7) }
      }
    }
  end
end

run(Lookup.new, title: "lookup")

File: rakugan/demo/lookup.pl
# A hash on the app: reading with a fallback, asking whether a key is
# there, and adding one while the window is open.
use Rakugan;

class Lookup {
    use Rakugan;
    field %prices = (apple => 120, banana => 80);
    field $picked = 0;
    field $label  = "none";

    method pick_apple {
        $picked = $prices{"apple"} // -1;
        $label = exists $prices{"cherry"} ? "cherry known" : "no cherry";
    }

    method add_cherry {
        $prices{"cherry"} = 200;
        $picked = $prices{"cherry"} // -1;
        $label = "cherry known" if exists $prices{"cherry"};
    }

    method view {
        return column(
            text("picked=$picked n=@{[ scalar keys %prices ]} $label"),
            text("apple costs @{[ $prices{'apple'} // -1 ]} right now", size => 12),
            row(
                button("apple",  on_click => sub { $self->pick_apple }),
                button("cherry", on_click => sub { $self->add_cherry }),
                button("miss",   on_click => sub { $picked = $prices{"durian"} // -7 }),
                spacing => 6,
            ),
            spacing => 8,
            padding => 12,
        );
    }
}

run(Lookup->new, title => "lookup");
